File: FastMines_uwp/src/fmg/uwp/app/Serializers/MosaicBackupDataSerializer.cs
using System;
using System.Linq;
using System.Collections.Generic;
using Windows.Foundation.Collections;
using Fmg.Common;
using Fmg.Common.Geom;
using Fmg.Core.App.Model;
using Fmg.Core.App.Serializers;
using Fmg.Core.Types;
using Fmg.Core.Mosaic.Cells;

namespace Fmg.Uwp.App.Serializers {

    /// <summary> Mosaic cells data (de)serializer. For save/restore <see cref="MosaicBackupData"/> </summary>
    public class MosaicBackupDataSerializer : ISerializer {

        private const int VERSION = 1;

        private const           string KEY__MOSAIC_BACKUP_DATA__SECTION             = nameof(MosaicBackupData);
        private const           string KEY__MOSAIC_BACKUP_DATA__VERSION             = "Version";
        private static readonly string KEY__MOSAIC_BACKUP_DATA__SIZE_FIELD_M        = nameof(MosaicBackupData.SizeField) + '.' + nameof(MosaicBackupData.SizeField.m);
        private static readonly string KEY__MOSAIC_BACKUP_DATA__SIZE_FIELD_N        = nameof(MosaicBackupData.SizeField) + '.' + nameof(MosaicBackupData.SizeField.n);
        private const           string KEY__MOSAIC_BACKUP_DATA__MOSAIC_TYPE         = nameof(MosaicBackupData.MosaicType);
        private static readonly string KEY__MOSAIC_BACKUP_DATA__CELL_STATES__STATUS = nameof(MosaicBackupData.CellStates) + '.' + nameof(BaseCell.StateCell.Status);
        private static readonly string KEY__MOSAIC_BACKUP_DATA__CELL_STATES__OPEN   = nameof(MosaicBackupData.CellStates) + '.' + nameof(BaseCell.StateCell.Open);
        private static readonly string KEY__MOSAIC_BACKUP_DATA__CELL_STATES__CLOSE  = nameof(MosaicBackupData.CellStates) + '.' + nameof(BaseCell.StateCell.Close);
        private static readonly string KEY__MOSAIC_BACKUP_DATA__CELL_STATES__DOWN   = nameof(MosaicBackupData.CellStates) + '.' + nameof(BaseCell.StateCell.Down);
        private const           string KEY__MOSAIC_BACKUP_DATA__AREA                = nameof(MosaicBackupData.Area);
        private const           string KEY__MOSAIC_BACKUP_DATA__CLICK_COUNT         = nameof(MosaicBackupData.ClickCount);

        internal MosaicBackupData Load(IPropertySet from) {
            if ((from != null) && from.ContainsKey(KEY__MOSAIC_BACKUP_DATA__SECTION)) try {
                var section = (Windows.Storage.ApplicationDataCompositeValue)from[KEY__MOSAIC_BACKUP_DATA__SECTION];
                if (!section.ContainsKey(KEY__MOSAIC_BACKUP_DATA__VERSION))
                    return null;

                int version = (int)section[KEY__MOSAIC_BACKUP_DATA__VERSION];
                if (version != 1)
                    throw new ArgumentException("MosaicBackupDataSerializer: Version #" + version + " is not supported");

                int mosaicOrdinal = (int)section[KEY__MOSAIC_BACKUP_DATA__MOSAIC_TYPE];
                int sizeFieldM    = (int)section[KEY__MOSAIC_BACKUP_DATA__SIZE_FIELD_M];
                int sizeFieldN    = (int)section[KEY__MOSAIC_BACKUP_DATA__SIZE_FIELD_N];
                double area    = (double)section[KEY__MOSAIC_BACKUP_DATA__AREA];
                int clickCount    = (int)section[KEY__MOSAIC_BACKUP_DATA__CLICK_COUNT];

                IList<EState> states = ((string)section[KEY__MOSAIC_BACKUP_DATA__CELL_STATES__STATUS]).Split(',').Select(v => Enum.Parse<EState>(v)).ToList();
                IList<EOpen > opens  = ((string)section[KEY__MOSAIC_BACKUP_DATA__CELL_STATES__OPEN  ]).Split(',').Select(v => Enum.Parse<EOpen >(v)).ToList();
                IList<EClose> closes = ((string)section[KEY__MOSAIC_BACKUP_DATA__CELL_STATES__CLOSE ]).Split(',').Select(v => Enum.Parse<EClose>(v)).ToList();
                IList< bool > downs  = ((string)section[KEY__MOSAIC_BACKUP_DATA__CELL_STATES__DOWN  ]).Split(',').Select(v => bool.Parse        (v)).ToList();

                System.Diagnostics.Debug.Assert(states.Count == sizeFieldM * sizeFieldN);
                System.Diagnostics.Debug.Assert(states.Count == opens.Count);
                System.Diagnostics.Debug.Assert(states.Count == closes.Count);

                MosaicBackupData data = new MosaicBackupData {
                    MosaicType = EMosaicEx.FromOrdinal(mosaicOrdinal),
                    SizeField  = new Matrisize(sizeFieldM, sizeFieldN),
                    Area       = area,
                    ClickCount = clickCount,
                    CellStates = new List<BaseCell.StateCell>(states.Count)
                };
                for (int i = 0; i < states.Count; ++i) {
                    data.CellStates.Add(new BaseCell.StateCell {
                        Status = states[i],
                        Open   = opens [i],
                        Close  = closes[i],
                        Down   = downs [i]
                    });
                }
                return data;

            } catch(Exception ex) {
                Logger.Error("Can`t load mosaic backup data from SharedPreferences" + ex);
            }
            return null;
        }

        internal void Save(MosaicBackupData data, IPropertySet to) {
            if (data == null)
                return;

            to[KEY__MOSAIC_BACKUP_DATA__SECTION] = new Windows.Storage.ApplicationDataCompositeValue() {
                [KEY__MOSAIC_BACKUP_DATA__VERSION            ] = VERSION,
                [KEY__MOSAIC_BACKUP_DATA__MOSAIC_TYPE        ] = data.MosaicType.Ordinal(),
                [KEY__MOSAIC_BACKUP_DATA__SIZE_FIELD_M       ] = data.SizeField.m,
                [KEY__MOSAIC_BACKUP_DATA__SIZE_FIELD_N       ] = data.SizeField.n,
                [KEY__MOSAIC_BACKUP_DATA__CELL_STATES__STATUS] = string.Join(",", data.CellStates.Select(c => c.Status.ToString())),
                [KEY__MOSAIC_BACKUP_DATA__CELL_STATES__OPEN  ] = string.Join(",", data.CellStates.Select(c => c.Open  .ToString())),
                [KEY__MOSAIC_BACKUP_DATA__CELL_STATES__CLOSE ] = string.Join(",", data.CellStates.Select(c => c.Close .ToString())),
                [KEY__MOSAIC_BACKUP_DATA__CELL_STATES__DOWN  ] = string.Join(",", data.CellStates.Select(c => c.Down  .ToString())),
                [KEY__MOSAIC_BACKUP_DATA__AREA               ] = data.Area,
                [KEY__MOSAIC_BACKUP_DATA__CLICK_COUNT        ] = data.ClickCount
            };
        }

    }

}

File: FastMines_android/src/main/java/fmg/android/app/serializers/MosaicBackupDataSerializer.java
package fmg.android.app.serializers;

import android.content.SharedPreferences;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import fmg.common.Logger;
import fmg.common.geom.Matrisize;
import fmg.core.app.model.MosaicBackupData;
import fmg.core.app.model.MosaicInitData;
import fmg.core.app.serializers.ISerializer;
import fmg.core.mosaic.cells.BaseCell;
import fmg.core.types.EClose;
import fmg.core.types.EMosaic;
import fmg.core.types.EOpen;
import fmg.core.types.EState;

/** Mosaic cells data (de)serializer. For save/restore {@link MosaicBackupData} */
public class MosaicBackupDataSerializer implements ISerializer {

    private static final int VERSION = 1;

    private static final String KEY__MOSAIC_BACKUP_DATA = MosaicBackupData.class.getSimpleName();
    private static final String KEY__MOSAIC_BACKUP_DATA__VERSION             = KEY__MOSAIC_BACKUP_DATA + ".version";
    private static final String KEY__MOSAIC_BACKUP_DATA__SIZE_FIELD_M        = KEY__MOSAIC_BACKUP_DATA + ".sizeField.M";
    private static final String KEY__MOSAIC_BACKUP_DATA__SIZE_FIELD_N        = KEY__MOSAIC_BACKUP_DATA + ".sizeField.N";
    private static final String KEY__MOSAIC_BACKUP_DATA__MOSAIC_TYPE         = KEY__MOSAIC_BACKUP_DATA + ".mosaicType";
    private static final String KEY__MOSAIC_BACKUP_DATA__CELL_STATES__STATUS = KEY__MOSAIC_BACKUP_DATA + ".cellStates.status";
    private static final String KEY__MOSAIC_BACKUP_DATA__CELL_STATES__OPEN   = KEY__MOSAIC_BACKUP_DATA + ".cellStates.open";
    private static final String KEY__MOSAIC_BACKUP_DATA__CELL_STATES__CLOSE  = KEY__MOSAIC_BACKUP_DATA + ".cellStates.close";
    private static final String KEY__MOSAIC_BACKUP_DATA__CELL_STATES__DOWN   = KEY__MOSAIC_BACKUP_DATA + ".cellStates.down";
    private static final String KEY__MOSAIC_BACKUP_DATA__AREA                = KEY__MOSAIC_BACKUP_DATA + ".area";
    private static final String KEY__MOSAIC_BACKUP_DATA__CLICK_COUNT         = KEY__MOSAIC_BACKUP_DATA + ".clickCount";

    MosaicBackupData load(SharedPreferences from) {
        if (from != null) try {
            MosaicBackupData data = new MosaicBackupData();
            int version = from.getInt(KEY__MOSAIC_BACKUP_DATA__VERSION, 0);
            if (version == 0)
                return null;
            if (version != 1)
                throw new IllegalArgumentException("MosaicBackupDataSerializer: Version #" + version + " is not supported");

            data.mosaicType = EMosaic.fromOrdinal(from.getInt  (KEY__MOSAIC_BACKUP_DATA__MOSAIC_TYPE , MosaicInitData.DEFAULT_MOSAIC_TYPE.ordinal()));
            data.sizeField = new Matrisize(       from.getInt  (KEY__MOSAIC_BACKUP_DATA__SIZE_FIELD_M, MosaicInitData.DEFAULT_SIZE_FIELD_M),
                                                  from.getInt  (KEY__MOSAIC_BACKUP_DATA__SIZE_FIELD_N, MosaicInitData.DEFAULT_SIZE_FIELD_M));
            data.area =                           from.getFloat(KEY__MOSAIC_BACKUP_DATA__AREA        , (float)MosaicInitData.AREA_MINIMUM);
            data.clickCount =                     from.getInt  (KEY__MOSAIC_BACKUP_DATA__CLICK_COUNT , 0);

            List<EState> states = Stream.of(from.getString(KEY__MOSAIC_BACKUP_DATA__CELL_STATES__STATUS, "").split(",")).map(EState ::valueOf).collect(Collectors.toList());
            List<EOpen > opens  = Stream.of(from.getString(KEY__MOSAIC_BACKUP_DATA__CELL_STATES__OPEN  , "").split(",")).map(EOpen  ::valueOf).collect(Collectors.toList());
            List<EClose> closes = Stream.of(from.getString(KEY__MOSAIC_BACKUP_DATA__CELL_STATES__CLOSE , "").split(",")).map(EClose ::valueOf).collect(Collectors.toList());
            List<Boolean> downs = Stream.of(from.getString(KEY__MOSAIC_BACKUP_DATA__CELL_STATES__DOWN  , "").split(",")).map(Boolean::valueOf).collect(Collectors.toList());

            assert (states.size() == data.sizeField.m * data.sizeField.n);
            assert (states.size() == opens.size());
            assert (states.size() == closes.size());

            data.cellStates = new ArrayList<>(states.size());
            for (int i = 0; i < states.size(); ++i) {
                BaseCell.StateCell stateCell = new BaseCell.StateCell();
                stateCell.setStatus(states.get(i));
                stateCell.setOpen(  opens .get(i));
                stateCell.setClose( closes.get(i));
                stateCell.setDown(  downs .get(i));
                data.cellStates.add(stateCell);
            }

            return data;

        } catch(Exception ex) {
            Logger.error("Can`t load mosaic backup data from SharedPreferences", ex);
        }
        return null;
    }

    void save(MosaicBackupData data, SharedPreferences.Editor editor) {
        if (data == null)
            return;
        editor.putInt(   KEY__MOSAIC_BACKUP_DATA__VERSION     , VERSION);
        editor.putInt(   KEY__MOSAIC_BACKUP_DATA__MOSAIC_TYPE , data.mosaicType.ordinal());
        editor.putInt(   KEY__MOSAIC_BACKUP_DATA__SIZE_FIELD_M, data.sizeField.m);
        editor.putInt(   KEY__MOSAIC_BACKUP_DATA__SIZE_FIELD_N, data.sizeField.n);
        editor.putString(KEY__MOSAIC_BACKUP_DATA__CELL_STATES__STATUS, data.cellStates.stream().map(c -> c.getStatus().name()).collect(Collectors.joining(",")));
        editor.putString(KEY__MOSAIC_BACKUP_DATA__CELL_STATES__OPEN  , data.cellStates.stream().map(c -> c.getOpen  ().name()).collect(Collectors.joining(",")));
        editor.putString(KEY__MOSAIC_BACKUP_DATA__CELL_STATES__CLOSE , data.cellStates.stream().map(c -> c.getClose ().name()).collect(Collectors.joining(",")));
        editor.putString(KEY__MOSAIC_BACKUP_DATA__CELL_STATES__DOWN  , data.cellStates.stream().map(c -> Boolean.toString(c.isDown())).collect(Collectors.joining(",")));
        editor.putFloat( KEY__MOSAIC_BACKUP_DATA__AREA        , (float)data.area);
        editor.putInt(   KEY__MOSAIC_BACKUP_DATA__CLICK_COUNT , data.clickCount);
    }

}